File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.java
package uk.gov.justice.digital.hmpps.oauth2server.verify;

import com.microsoft.applicationinsights.TelemetryClient;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository;

import java.util.Map;
import java.util.Optional;

@Service
@Slf4j
@Transactional(transactionManager = "authTransactionManager", readOnly = true)
public class TokenService {

    private final UserTokenRepository userTokenRepository;
    private final TelemetryClient telemetryClient;

    public TokenService(final UserTokenRepository userTokenRepository,
                        final TelemetryClient telemetryClient) {
        this.userTokenRepository = userTokenRepository;
        this.telemetryClient = telemetryClient;
    }

    public Optional<UserToken> getToken(final TokenType tokenType, final String token) {
        final var userTokenOptional = userTokenRepository.findById(token);
        return userTokenOptional.filter(t -> t.getTokenType() == tokenType);
    }

    public Optional<String> checkToken(final TokenType tokenType, final String token) {
        final var userTokenOptional = getToken(tokenType, token);
        if (userTokenOptional.isEmpty()) {
            log.info("Failed to {} password due to invalid token", tokenType.getDescription());
            telemetryClient.trackEvent(String.format("%sPasswordFailure", tokenType.getDescription()),
                    Map.of("reason", "invalid"), null);
            return Optional.of("invalid");
        }
        final var userToken = userTokenOptional.get();
        final var username = userToken.getUser().getUsername();
        if (userToken.hasTokenExpired()) {
            log.info("Failed to reset password due to expired token");
            telemetryClient.trackEvent(String.format("%sPasswordFailure", tokenType.getDescription()),
                    Map.of("username", username, "reason", "expired"), null);
            return Optional.of("expired");
        }
        return Optional.empty();
    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/verify/TokenService.kt
package uk.gov.justice.digital.hmpps.oauth2server.verify

import com.microsoft.applicationinsights.TelemetryClient
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository
import uk.gov.justice.digital.hmpps.oauth2server.security.UserService
import java.util.*
import javax.persistence.EntityNotFoundException

@Service
@Transactional(transactionManager = "authTransactionManager", readOnly = true)
class TokenService(private val userTokenRepository: UserTokenRepository,
                   private val userService: UserService,
                   private val telemetryClient: TelemetryClient) {

  companion object {
    val log: Logger = LoggerFactory.getLogger(this::class.java)
  }

  fun getToken(tokenType: TokenType, token: String): Optional<UserToken> {
    val userTokenOptional = userTokenRepository.findById(token)
    return userTokenOptional.filter { t -> t.tokenType == tokenType }
  }

  fun getUserFromToken(tokenType: TokenType, token: String): User {
    val userTokenOptional = userTokenRepository.findById(token)
    val userToken = userTokenOptional.filter { t -> t.tokenType == tokenType }.orElseThrow { EntityNotFoundException("Token not found $token") }
    return userToken.user
  }

  fun checkToken(tokenType: TokenType, token: String): Optional<String> {
    val userTokenOptional = getToken(tokenType, token)
    if (userTokenOptional.isEmpty) {
      log.info("Failed to {} due to invalid token", tokenType.description)
      telemetryClient.trackEvent("${tokenType.description}Failure",
          mapOf("reason" to "invalid"), null)
      return Optional.of("invalid")
    }
    val userToken = userTokenOptional.get()
    if (userToken.hasTokenExpired()) {
      log.info("Failed to {} due to expired token", tokenType.description)
      val username = userToken.user.username
      telemetryClient.trackEvent("${tokenType.description}Failure",
          mapOf("username" to username, "reason" to "expired"), null)
      return Optional.of("expired")
    }
    return Optional.empty()
  }

  @Transactional(transactionManager = "authTransactionManager")
  fun createToken(tokenType: TokenType, username: String): String {
    log.info("Requesting {} for {}", tokenType.description, username)
    val user = userService.getOrCreateUser(username)
    val userToken = user.createToken(tokenType)
    telemetryClient.trackEvent("${tokenType.description}Request",
        mapOf("username" to username), null)
    return userToken.token
  }

  @Transactional(transactionManager = "authTransactionManager")
  fun removeToken(tokenType: TokenType, token: String) =
      getToken(tokenType, token).ifPresent { userTokenRepository.delete(it) }
}
